freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/FreeIpaClientFactory.java

```diff
@@ -70,42 +73,48 @@ private String toClusterProxyBasepath(String freeIpaClusterCrn) {
         return String.format("%s%s", clusterProxyService.getProxyPath(freeIpaClusterCrn), FreeIpaClientBuilder.DEFAULT_BASE_PATH);
     }
 
-    public FreeIpaClient getFreeIpaClientForStack(Stack stack) throws FreeIpaClientException {
-        LOGGER.debug("Creating FreeIpaClient for stack {}", stack.getResourceCrn());
+    private FreeIpaClient getFreeIpaClient(Stack stack, boolean withPing) throws FreeIpaClientException {
+        stack = stackService.getByIdWithListsInTransaction(stack.getId());
         Status stackStatus = stack.getStackStatus().getStatus();
         if (!stackStatus.isFreeIpaUnreachableStatus()) {
             try {
                 if (clusterProxyService.isCreateConfigForClusterProxy(stack)) {
-                    return getFreeIpaClientBuilderForClusterProxy(stack).build();
+                    return getFreeIpaClientBuilderForClusterProxy(stack).build(withPing);
                 } else {
-                    return getFreeIpaClientBuilder(stack).build();
+                    Exception lastException = new FreeIpaClientException("No instances exist in the stack.");
```

Comment:
this exception driven logic and return inside the for loop makes the code hard to read.
I think it would be more readable if you do something like have an `Optional<FreeIpaClient> client = Optional.empty()` and iterate through the instances while it's empty and you have more instances. In the end you could return with the client or throw an exception

Reply:
changed.